examples/ssl_metrics.py

```diff
@@ -0,0 +1,211 @@
+        transforms.ToImage(),
+        transforms.ToDtype(torch.float32, scale=True),
+    ])
+
+    ds_builder = load_dataset_builder(args.dataset, trust_remote_code=True)
+    ds_builder.download_and_prepare()
+
+    dataset = ds_builder.as_dataset(split='train')
+    dataset = dataset.to_iterable_dataset(num_shards=world_size * max(args.workers, 1))
+    dataset = split_dataset_by_node(dataset, rank=rank, world_size=world_size)
+    dataset = dataset.map(ssl_augment(transform_reg, transform_aug, args.q))
+    loader = DataLoader(dataset, batch_size=8, pin_memory=True, num_workers=args.workers)
+
+    rank_me_embeddings = []
+    lidar_embeddings = []
+
+    n_iter = int(math.ceil(args.n / world_size))
+    for i, batch in tqdm(enumerate(loader), total=n_iter, disable=rank > 0):
+        if i == n_iter:
+            break
+
+        images = batch['image'].cuda(non_blocking=True)
+        # Roll the batch and variant dimensions into batch
+        flat_images = images.flatten(0, 1)
+
+        with torch.autocast(device_type=device.type, dtype=torch.bfloat16):
+            summary, features = model(flat_images)
+
+        # Unroll the batch and variant dimensions
+        g_summary = summary.reshape(*images.shape[:2], *summary.shape[1:])
+        g_features = features.reshape(*images.shape[:2], *features.shape[1:])
+
+        # RankMe is only computed for the regular images
+        rank_me_embeddings.append(g_summary[:, 0])
+        lidar_embeddings.append(g_summary)
+
+    rank_me_embeddings = gather_cat(torch.cat(rank_me_embeddings), rank, world_size)
+    lidar_embeddings = gather_cat(torch.cat(lidar_embeddings), rank, world_size)
+
+    if rank > 0:
+        return
+
+    rank_me = calc_rank_me(rank_me_embeddings)
+    lidar = calc_lidar(lidar_embeddings)
+
+    print(f'RankMe: {rank_me.item():.3f}')
+    print(f'LiDAR: {lidar.item():.3f}')
+
+
+def ssl_augment(transform_reg, transform_aug, num_variants):
+    def augment(ex: Dict[str, torch.Tensor]):
+        image: Image.Image = ex['image']
+        image = image.convert('RGB')
+        im_reg = transform_reg(image)
+        variants = [transform_aug(image) for _ in range(num_variants)]
+        variants.insert(0, im_reg)
+
+        ret = torch.stack(variants)
+        return dict(image=ret)
+    return augment
+
+
+def calc_rank_me(embeddings: torch.Tensor):
+    # embeddings: N,K
+
+    eig = torch.linalg.svdvals(embeddings)
+
+    return entropy_rank(eig)
+
+
+def calc_lidar(embeddings: torch.Tensor, delta: float = 1e-5):
+    # embeddings: N,V,K
+    embeddings = embeddings.double()
+
+    mu_x = embeddings.mean(dim=1)  # N,K
+    mu = mu_x.mean(dim=0, keepdim=True)  # 1,K
+
+    mu_zc = mu_x - mu
+    cov_b = (mu_zc.T @ mu_zc) / (mu_zc.shape[0] - 1)
+
+    ex_zc = (embeddings - mu_x.unsqueeze(1)).flatten(0, 1)
+    cov_w = (ex_zc.T @ ex_zc) / (ex_zc.shape[0] - 1)
+    cov_w.diagonal().add_(delta)
+
+    e_cov_w, v_cov_w = torch.linalg.eigh(cov_w)
```

Comment:
Why are we doing an eigen value decomposition here? I don't see this in the paper (equations 2 and 3?)

Reply:
It enables us to compute $\Sigma_w(e)^{\frac{1}{2}}$ (needed in equation 3). 

To compute a matrix power to a non-positive integer, you need to diagonalize the matrix (using `torch.linalg.eig`). Additionally, because our covariance matrix is positive definite, we can use `torch.linalg.eigh`. Once diagonalized, you can compute a matrix raised to an arbitrary power by raising the diagonal eigenvalue matrix to that power, and then re-composing the matrix, as we do on line 171.

Writing all of this, I realize that my in-code comments are... lacking 😅

Reply:
Thanks!